### PR TITLE
process cycles as soon as they are detected

### LINKED ISSUE #32326: Invalid Recursive Enum (2+ occurances) causes compiler out-of-memory on beta/nightly
For example, compiling

```rust
enum Expr {
    Plus(Expr, Expr),
    Literal(i64),
}
```
 with `rustc` causes the error message `fatal runtime error: out of memory`

Single occurrences seem to work fine.

### PR DESCRIPTION
**Pull request by nikomatsakis**
We used to wait for the recursion limit, but that might well be too
long!

Fixes #32326 
r? @aturon

### CODE REVIEW

File: src/librustc/traits/fulfill.rs [hunk 1]
```diff
@@ -320,103 +320,172 @@ impl<'tcx> FulfillmentContext<'tcx> {
...
+            // ignore. If the obligation appears in our *ancestors*
+            // (rather than some more distant relative), that
+            // indicates a cycle. Cycles are either considered
+            // resolved (if this is a coinductive case) or a fatal
+            // error.
+            if let Some(index) = ancestor_set.has(selcx.infcx(), &obligation.predicate) {
+                //                            ~~~ (*) see above
+                debug!("process_child_obligations: cycle index = {}", index);
+
+                if coinductive_match(selcx, &obligation, &backtrace) {
```
Comment (nikomatsakis) [inline src/librustc/traits/fulfill.rs]: just realized that this is a bug -- we are overlooking the `pending_obligation`, which is at the logical "top" of the backtrace here

File: src/librustc/traits/fulfill.rs [hunk 2]
```diff
@@ -320,103 +320,172 @@ impl<'tcx> FulfillmentContext<'tcx> {
...
+        if !self.populated {
+            let backtrace = self.backtrace.clone();
+            for (index, ancestor) in backtrace.enumerate() {
+                // Ugh. This just feels ridiculously
+                // inefficient.  But we need to compare
+                // predicates without being concerned about
+                // the vagaries of type inference, so for now
+                // just ensure that they are always
+                // up-to-date. (I suppose we could just use a
+                // snapshot and check if they are unifiable?)
```
Comment (arielb1) [inline src/librustc/traits/fulfill.rs]: Isn't resolving type variables *once* enough, given that variable resolution is well-founded? at worst we will explore a branch a bit too deeply.

Comment (nikomatsakis) [inline src/librustc/traits/fulfill.rs]: @arielb1 

> Isn't resolving type variables once enough, given that variable resolution is well-founded? at worst we will explore a branch a bit too deeply.

I don't understand what you mean. Maybe you can elaborate?

My concern is basically that if we have any types in the backtrace that are out of date, then the hashtable will not find them, and we will mistake a cycle for a duplicate obligation.

Comment (arielb1) [inline src/librustc/traits/fulfill.rs]: Mistake a cycle for a duplicate obligation? If the duplicate set itself is never resolved, then `is_duplicate_or_add` and `is_ancestor` do both the same check, so there should be no problem.

Comment (arielb1) [inline src/librustc/traits/fulfill.rs]: This obligation forest code is so terribly complicated.

**Comment by aturon** [comment]
@nikomatsakis OK, I've reviewed this and it's a clear improvement. r=me after a rebase.

**Pull request merged**